modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4SizeHeaderFrameDecoder.java

```diff
@@ -35,7 +35,7 @@
     @Override
     protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
         try {
-            boolean continueDecode = true;
+            boolean continueDecode = in.readableBytes() >= HEADER_SIZE;
             while (continueDecode) {
                 int messageLength = TcpTransport.readMessageLength(Netty4Utils.toBytesReference(in));
```

Comment:
At the cost of some duplication (between Netty and NIO impls.) we could just remove creating the wrapping `BytesReference` just for reading the length here as well. I'd look into that in a follow up.